src/stan_math_backend/Cpp.ml

```diff
@@ -0,0 +1,732 @@
+
+  (** Set up the try/catch logic for throwing an exception with
+      its location set to the Stan program location. *)
+  let rethrow_located stmts =
+    TryCatch
+      ( block stmts
+      , (Types.const_ref (Type_literal "std::exception"), "e")
+      , Block
+          [ Expression
+              (FunCall
+                 ( "stan::lang::rethrow_located"
+                 , []
+                 , [ Var "e"
+                   ; Index (Var "locations_array__", Var "current_statement__")
+                   ] ) ) ] )
+
+  let fori loopvar lower upper body =
+    let init =
+      make_var_defn ~type_:Int ~name:loopvar ~init:(Assignment lower) () in
+    let stop = BinOp (Var loopvar, LEq, upper) in
+    let incr = Unary (Incr, Var loopvar) in
+    For (init, stop, incr, body)
+
+  let if_block cond stmts = IfElse (cond, block stmts, None)
+
+  (** Supress warnings for a variable which may not be used. *)
+  let unused s =
+    [Comment "supress unused var warning"; Expression (Cast (Void, Var s))]
+end
+
+module Decls = struct
+  (** Declarations which get re-used often in the generated model *)
+
+  let current_statement =
+    VarDef
+      (make_var_defn ~type_:Int ~name:"current_statement__"
+         ~init:(Assignment (Literal "0")) () )
+
+  let dummy_var =
+    VarDef
+      (make_var_defn ~type_:Types.local_scalar ~name:"DUMMY_VAR__"
+         ~init:(Construction [Exprs.quiet_NaN])
+         () )
+    :: Stmts.unused "DUMMY_VAR__"
+
+  let serializer_in =
+    VarDef
+      (make_var_defn
+         ~type_:(TypeTrait ("stan::io::deserializer", [Types.local_scalar]))
+         ~name:"in__"
+         ~init:(Construction [Var "params_r__"; Var "params_i__"])
+         () )
+
+  let serializer_out =
+    VarDef
+      (make_var_defn
+         ~type_:(TypeTrait ("stan::io::serializer", [Types.local_scalar]))
+         ~name:"out__"
+         ~init:(Construction [Var "vars__"])
+         () )
+
+  let lp_accum t =
+    VarDef
+      (make_var_defn
+         ~type_:(TypeTrait ("stan::math::accumulator", [t]))
+         ~name:"lp_accum__" () )
+end
+
+type template_parameter =
+  | Typename of string  (** The name of a template typename *)
+  | RequireIs of string * string
+      (** A C++ type trait (e.g. is_arithmetic) and the template
+          name which needs to satisfy that.
+          These are collated into one require_all_t<> *)
+  | Require of string * string list
+  | Bool of string  (** A named boolean template type *)
+[@@deriving sexp]
+
+type cv_qualifiers = Const | Final | NoExcept [@@deriving sexp]
+
+type fun_defn =
+  { templates_init: template_parameter list list * bool [@default [[]], false]
+        (** Double list since some functions (mainly reduce_sum functors) need two sets of templates *)
+  ; name: identifier
+  ; inline: bool [@default false]
+  ; return_type: type_
+  ; args: (type_ * string) list
+  ; cv_qualifiers: cv_qualifiers list [@default []]
+  ; body: stmt list option }
+[@@deriving make, sexp]
```

Comment:
Do you mind shifting these around to follow the standard syntax scheme of a C++ function?
```suggestion
type fun_defn =
  { templates_init: template_parameter list list * bool [@default [[]], false]
        (** Double list since some functions (mainly reduce_sum functors) need two sets of templates *)
  ; inline: bool [@default false]
  ; return_type: type_
  ; name: identifier
  ; args: (type_ * string) list
  ; cv_qualifiers: cv_qualifiers list [@default []]
  ; body: stmt list option }
[@@deriving make, sexp]
```

I would be nice to use the same verbage as the standard for each of these parts of the function decleration

https://en.cppreference.com/w/cpp/language/function